File: domain/learner-model.ts
import {
  CONFIDENCE_WINDOW,
  MASTERY_DIFFICULTY_WEIGHT,
  MASTERY_FAIL_MULTIPLIER,
  MASTERY_PARTIAL_FACTOR,
  MASTERY_REPEAT_DAMPING,
  MASTERY_SUCCESS_GROWTH,
  NEXT_REVIEW_STABILITY_FRACTION,
  STATUS_FADING_GAP,
  STATUS_FADING_MASTERY,
  STATUS_STRONG_MASTERY,
  STATUS_STRONG_RETENTION,
  STATUS_WEAK_MASTERY,
  STATUS_WEAK_RETENTION,
  STABILITY_BASE_DAYS,
  STABILITY_MASTERY_WEIGHT,
} from "./constants";
import type { Concept, ConceptStatus, LearningEvent, RetrievalOutcome } from "./types";

export function clamp01(value: number) {
  if (!Number.isFinite(value)) return 0;
  return Math.min(1, Math.max(0, value));
}

export function daysBetween(fromIso: string, toIso: string) {
  return Math.max(0, (Date.parse(toIso) - Date.parse(fromIso)) / 86_400_000);
}

export function stabilityDays(mastery: number, successfulRetrievals: number) {
  return STABILITY_BASE_DAYS + STABILITY_MASTERY_WEIGHT * clamp01(mastery) * Math.log(1 + Math.max(0, successfulRetrievals));
}

export function predictedRetention(mastery: number, successfulRetrievals: number, lastSuccessAt: string | null, nowIso: string) {
  if (!lastSuccessAt) return 0;
  const stability = stabilityDays(mastery, successfulRetrievals);
  const elapsed = daysBetween(lastSuccessAt, nowIso);
  return clamp01(mastery * 2 ** (-elapsed / stability));
}

export function nextReviewAt(mastery: number, successfulRetrievals: number, lastSuccessAt: string | null, nowIso: string) {
  if (!lastSuccessAt) return nowIso;
  const days = stabilityDays(mastery, successfulRetrievals) * NEXT_REVIEW_STABILITY_FRACTION;
  return new Date(Date.parse(nowIso) + days * 86_400_000).toISOString();
}

export function deriveStatus(mastery: number, retention: number, retrievalAttempts: number): ConceptStatus {
  if (retrievalAttempts <= 0) return "not_learned";
  if (mastery < STATUS_WEAK_MASTERY || retention < STATUS_WEAK_RETENTION) return "weak";
  if (mastery >= STATUS_STRONG_MASTERY && retention >= STATUS_STRONG_RETENTION) return "strong";
  if (mastery >= STATUS_FADING_MASTERY && retention <= mastery - STATUS_FADING_GAP) return "fading";
  return "stable";
}

export function applyRetrievalMastery(mastery: number, outcome: RetrievalOutcome, difficulty: number, successfulRetrievals: number) {
  const current = clamp01(mastery);
  if (outcome === "failure") return clamp01(current * MASTERY_FAIL_MULTIPLIER);
  const growth = MASTERY_SUCCESS_GROWTH
    * (1 - MASTERY_DIFFICULTY_WEIGHT * clamp01(difficulty))
    / (1 + MASTERY_REPEAT_DAMPING * Math.max(0, successfulRetrievals));
  const scaled = outcome === "partial" ? growth * MASTERY_PARTIAL_FACTOR : growth;
  return clamp01(current + (1 - current) * scaled);
}

export function confidenceFromOutcomes(outcomes: RetrievalOutcome[]) {
  const recent = outcomes.slice(-CONFIDENCE_WINDOW);
  if (!recent.length) return 0;
  const score = recent.reduce((total, outcome) => total + (outcome === "success" ? 1 : outcome === "partial" ? 0.5 : 0), 0);
  return clamp01(score / recent.length);
}

export type ConceptCache = Pick<Concept,
  | "mastery"
  | "confidence"
  | "predictedRetention"
  | "lastReviewedAt"
  | "nextReviewAt"
  | "retrievalAttempts"
  | "successfulRetrievals"
  | "failedRetrievals"
  | "updatedAt"
>;

function eventTime(event: LearningEvent) {
  return Date.parse(event.createdAt);
}

export function recomputeConceptCache(
  concept: Pick<Concept, "id" | "difficulty">,
  events: LearningEvent[],
  nowIso: string,
): ConceptCache {
  const history = events
    .filter((event) => event.conceptId === concept.id)
    .sort((a, b) => eventTime(a) - eventTime(b) || a.id.localeCompare(b.id));

  let mastery = 0;
  let retrievalAttempts = 0;
  let successfulRetrievals = 0;
  let failedRetrievals = 0;
  let lastReviewedAt: string | null = null;
  let lastSuccessAt: string | null = null;
  const retrievalOutcomes: RetrievalOutcome[] = [];

  for (const event of history) {
    if (event.kind === "seed_rating") {
      mastery = event.masteryAfter;
      lastReviewedAt = event.createdAt;
      if (event.outcome === "success") lastSuccessAt = event.createdAt;
      continue;
    }
    mastery = applyRetrievalMastery(mastery, event.outcome, concept.difficulty, successfulRetrievals);
    retrievalAttempts += 1;
    lastReviewedAt = event.createdAt;
    retrievalOutcomes.push(event.outcome);
    if (event.outcome === "success") {
      successfulRetrievals += 1;
      lastSuccessAt = event.createdAt;
    } else if (event.outcome === "failure") {
      failedRetrievals += 1;
    } else {
      successfulRetrievals += 1;
      lastSuccessAt = event.createdAt;
    }
  }

  const retention = predictedRetention(mastery, successfulRetrievals, lastSuccessAt, nowIso);
  return {
    mastery,
    confidence: confidenceFromOutcomes(retrievalOutcomes),
    predictedRetention: retention,
    lastReviewedAt,
    nextReviewAt: nextReviewAt(mastery, successfulRetrievals, lastSuccessAt, nowIso),
    retrievalAttempts,
    successfulRetrievals,
    failedRetrievals,
    updatedAt: nowIso,
  };
}

export function withCachedState(concept: Concept, cache: ConceptCache): Concept {
  return { ...concept, ...cache };
}

File: ios/Domain/Sources/VellumDomain/LearnerModel.swift
import Foundation

public enum LearnerModel {
    public static func clamp01(_ value: Double) -> Double {
        guard value.isFinite else { return 0 }
        return min(1, max(0, value))
    }

    public static func daysBetween(_ from: Date, _ to: Date) -> Double {
        max(0, to.timeIntervalSince(from) / LearnerConstants.secondsPerDay)
    }

    public static func stabilityDays(mastery: Double, successfulRetrievals: Int) -> Double {
        LearnerConstants.stabilityBaseDays
            + LearnerConstants.stabilityMasteryWeight
            * clamp01(mastery)
            * log(1 + Double(max(0, successfulRetrievals)))
    }

    public static func applyRetrievalMastery(
        mastery: Double,
        outcome: RetrievalOutcome,
        difficulty: Double,
        successfulRetrievals: Int
    ) -> Double {
        let current = clamp01(mastery)
        if outcome == .failure {
            return clamp01(current * LearnerConstants.masteryFailMultiplier)
        }
        let growth = LearnerConstants.masterySuccessGrowth
            * (1 - LearnerConstants.masteryDifficultyWeight * clamp01(difficulty))
            / (1 + LearnerConstants.masteryRepeatDamping * Double(max(0, successfulRetrievals)))
        let scaled = outcome == .partial ? growth * LearnerConstants.masteryPartialFactor : growth
        return clamp01(current + (1 - current) * scaled)
    }

    public static func confidenceFromOutcomes(_ outcomes: [RetrievalOutcome]) -> Double {
        let recent = Array(outcomes.suffix(LearnerConstants.confidenceWindow))
        guard !recent.isEmpty else { return 0 }
        let score = recent.reduce(0.0) { total, outcome in
            switch outcome {
            case .success: total + 1
            case .partial: total + 0.5
            case .failure: total
            }
        }
        return clamp01(score / Double(recent.count))
    }

    public static func deriveStatus(
        mastery: Double,
        retention: Double,
        retrievalAttempts: Int
    ) -> ConceptStatus {
        if retrievalAttempts <= 0 && mastery <= 0 { return .notLearned }
        if mastery < LearnerConstants.statusWeakMastery || retention < LearnerConstants.statusWeakRetention {
            return .weak
        }
        if mastery >= LearnerConstants.statusStrongMastery
            && retention >= LearnerConstants.statusStrongRetention
        {
            return .strong
        }
        if mastery >= LearnerConstants.statusFadingMastery
            && retention <= mastery - LearnerConstants.statusFadingGap
        {
            return .fading
        }
        return .stable
    }

    public static func recomputeConceptCache(
        conceptId: String,
        difficulty: Double,
        events: [LearningEvent],
        now: Date
    ) -> ConceptCache {
        let history = events
            .filter { $0.conceptId == conceptId }
            .sorted {
                if $0.createdAt != $1.createdAt { return $0.createdAt < $1.createdAt }
                return $0.id < $1.id
            }

        var mastery = 0.0
        var retrievalAttempts = 0
        var successfulRetrievals = 0
        var failedRetrievals = 0
        var lastReviewedAt: Date?
        var lastSuccessAt: Date?
        var retrievalOutcomes: [RetrievalOutcome] = []

        for event in history {
            if event.kind == .seedRating {
                mastery = event.masteryAfter
                lastReviewedAt = event.createdAt
                if event.outcome == .success || event.outcome == .partial {
                    lastSuccessAt = event.createdAt
                    successfulRetrievals = 1
                }
                continue
            }
            mastery = applyRetrievalMastery(
                mastery: mastery,
                outcome: event.outcome,
                difficulty: difficulty,
                successfulRetrievals: successfulRetrievals
            )
            retrievalAttempts += 1
            lastReviewedAt = event.createdAt
            retrievalOutcomes.append(event.outcome)
            if event.outcome == .failure {
                failedRetrievals += 1
            } else {
                successfulRetrievals += 1
                lastSuccessAt = event.createdAt
            }
        }

        let retention = RetentionEngine.predictedRetention(
            mastery: mastery,
            successfulRetrievals: successfulRetrievals,
            lastSuccessAt: lastSuccessAt,
            now: now
        )
        return ConceptCache(
            mastery: mastery,
            confidence: confidenceFromOutcomes(retrievalOutcomes),
            predictedRetention: retention,
            lastReviewedAt: lastReviewedAt,
            nextReviewAt: RetentionEngine.nextReviewAt(
                mastery: mastery,
                successfulRetrievals: successfulRetrievals,
                lastSuccessAt: lastSuccessAt,
                now: now
            ),
            retrievalAttempts: retrievalAttempts,
            successfulRetrievals: successfulRetrievals,
            failedRetrievals: failedRetrievals
        )
    }

    public static func applying(_ cache: ConceptCache, to concept: Concept) -> Concept {
        var next = concept
        next.mastery = cache.mastery
        next.confidence = cache.confidence
        next.predictedRetention = cache.predictedRetention
        next.lastReviewedAt = cache.lastReviewedAt
        next.nextReviewAt = cache.nextReviewAt
        next.retrievalAttempts = cache.retrievalAttempts
        next.successfulRetrievals = cache.successfulRetrievals
        next.failedRetrievals = cache.failedRetrievals
        return next
    }

    public static func refreshConcepts(_ concepts: [Concept], events: [LearningEvent], now: Date) -> [Concept] {
        concepts.map { concept in
            applying(
                recomputeConceptCache(
                    conceptId: concept.id,
                    difficulty: concept.difficulty,
                    events: events,
                    now: now
                ),
                to: concept
            )
        }
    }
}
